clarify that the feature gate must be disabled by default for the dep… (#28734)
kubernetes/website: content/en/docs/reference/using-api/deprecation-policy.md

  * Alpha: the feature gate is disabled by default and can be enabled by the user.
  * Beta: the feature gate is enabled by default and can be disabled by the user.
  * Deprecated: the feature gate is disabled by default and can be enabled

>   by the user for the duration of deprecation window. See ["Deprecation"](#deprecation).

Reviewer: ```suggestion
  by the user for the duration of the deprecation window. See [Deprecation](#deprecation) for details.
```

>   * GA: the feature gate is deprecated and becomes non-operational - enabled by

Reviewer: ```suggestion
  * GA: the feature gate is deprecated and becomes non-operational. The feature is enabled by
```

  default and cannot be disabled by the user.

>   * GA or Deprecated, deprecation window complete: the feature gate is removed

Reviewer: ```suggestion
  * GA or Deprecated (deprecation window complete): the feature gate is removed
```

  * GA: the feature gate is deprecated (see ["Deprecation"](#deprecation)) and becomes
  non-operational.
  * GA, deprecation window complete: the feature gate is removed and calls to it are
  no longer accepted. Feature gate is disabled by default and can be enabled by

>   the user during the deprecation window.

Reviewer 2: I think this is wrong. One a feature is fully stable, you cannot set the feature gate  that previously controlled it. In fact, attempting to do so may be treated as an error.

In this phase the feature is moving from _stable and the feature gate setting is ignored_ to _stable and the feature gate setting is invalid_.

This text is not about deprecating stable features. If we did add a feature (let's say `WatchBookmark`), and then deprecated it, the `WatchBookmark` feature gate would stay removed. We would leave the watch bookmark behavior in the API but we'd deprecated it, separately, and might warn people when they specify query strings etc that involve it.

Possibly but unlikely: we might add a new feature `ForbidWatchBookmark` that when enabled always returns a 403 should someone try to use a watch bookmark. This kind of deprecation-and-feature-removal is so rare that I don't have an example.

Author: hm, GA'd feature gate cannot be deprecated as per the next section. I think this section calls removal of a feature gate and locking it to `true` with the term "deprecation". Where this deprecation is only deprecates the feature gate use. Which is very confusing as can be mistaken with the feature deprecation. Let me rephrase

**Rule #9: Deprecated feature gates must respond with a warning when used. When a feature gate
is deprecated it must be documented in both in the release notes and the corresponding CLI help.
Both warnings and documentation must indicate whether a feature gate is non-operational.

> Feature gate must be disabled by default for the deprecation window.**

Reviewer 2: ```suggestion
Throughout the deprecation period, the feature gate must be disabled by default.
```
BTW is this true? If we had a legacy-mode feature gate such as `LegacyNodeRoleBehavior` and then wanted to deprecate the feature gate, how would that work?

Reviewer 3: yeah, feature gates *can* be false by default during the deprecation window, but I don't think it's **required**

Author: if feature gate is still in code, it means deprecation happens from alpha or beta. If it is alpha, then nothing changes, feature gate keep being disabled. If it was in beta, what would be the reason to keep feature gate enabled by default? Since we want to discourage use of the feature, we want to make sure to not have it enabled by default on upgrade. So it will be an intentional step to re-enable it. I can change must to should if this is better.

As for `LegacyNodeRoleBehavior`, it is already GA. Deprecation is not possible after the GA as per text above in this section. So there will be a separate deprecation feature gate.

Reviewer 2: I wanted to suggest a thought experiment: imagine there is a brand new feature gate `LegacyPersistentVolumeBehavior', and like `LegacyNodeRoleBehavior` it starts life as alpha, but because of circumstances, we set `LegacyPersistentVolumeBehavior' true by default.

Then, as a project, we have a change of heart. We decide not to fix the imaginary defect in the PV API, and instead we'll live with its shortcomings. We deprecate `LegacyPersistentVolumeBehavior`, leaving it true. Until that deprecation is complete you can set it false and get the new, alpha behavior for PVs, but you're on marked time. We never default `LegacyPersistentVolumeBehavior` to false but we do eventually take that feature gate out of the CLI options for the binaries we ship and the code they build from.

When I think about this experiment, it makes me not want to say:
> Throughout the deprecation period, the feature gate must be disabled by default.

Author: as per previous section:

>  * Alpha: the feature gate is disabled by default and can be enabled by the user.
>  * Beta: the feature gate is enabled by default and can be disabled by the user.
>  * GA: the feature gate is deprecated (see ["Deprecation"](#deprecation)) and becomes
>  non-operational.
>  * GA, deprecation window complete: the feature gate is removed and calls to it are
>  no longer accepted.

I used disabled/enabled with the same meaning, not with the meaning of actual value being "true" or "false".

Reviewer 3: the natural way to read disabled/enabled is to map to false/true... I still don't think requiring a deprecated gate to be disabled/false is universally necessary

**Rule #9: Deprecated feature gates must respond with a warning when used. When a feature gate
is deprecated it must be documented in both in the release notes and the corresponding CLI help.
Both warnings and documentation must indicate whether a feature gate is non-operational.

> Throughout the deprecation period, the feature gate must be disabled by default.**

Reviewer: ```suggestion
Throughout the deprecation period, the feature gate should typically be disabled by default.**
```

Maybe "should typically" so it's a recommendation and not a rule?
